backend/xivdpscalc/pps/sample_sch_rotation.py

```diff
@@ -0,0 +1,81 @@
+
+    def get_action(self, current_time: ElapsedTime, cooldowns: dict[SchAction, float],
+                   remaining_effect_durations: dict[SchEffect, float], resources: dict[SchResource, int]) -> SchAction:
+        """
+        Returns a SchAction based on the fixed rotation
+        """
+        if self.index < len(self.rotation):
+            self.index += 1
+            return self.rotation[self.index - 1]
+        return self.default_action
+
+class SampleSchRotation(SchRotation): #pylint: disable=too-few-public-methods
+    """
+    An example rotation exibiting more dynamic decision rotation generation.
+    This is not meant to be a perfect replication of a good sch rotation, this is primarily for testing and demonstration.
+    This operates by starting with a gcd 4 chain opener, and then using a rudimentary priority list to pick actions.
+    """
+
+    def __init__(self):
+        self.rotation = [SchAction.BIOLYSIS, SchAction.AETHERFLOW, SchAction.ENERGYDRAIN,
+                         SchAction.BROIL3, SchAction.BROIL3, SchAction.RUIN2, SchAction.CHAINSTRATAGEM, SchAction.ENERGYDRAIN,
+                         SchAction.BROIL3, SchAction.RUIN2, SchAction.ENERGYDRAIN, SchAction.SWIFTCAST, SchAction.BROIL3,
+                         SchAction.DISSIPATION, SchAction.ENERGYDRAIN, SchAction.BROIL3, SchAction.RUIN2, SchAction.ENERGYDRAIN]
+        self.index = 0
+
+    def get_action(self, current_time: ElapsedTime, cooldowns: dict[SchAction, float],
+                   remaining_effect_durations: dict[SchEffect, float], resources: dict[SchResource, int]) -> SchAction:
+        """
+        Returns a SchAction based on an arbitrary set of rules.
+        """
+        # broil by default
+        selected_action = SchAction.BROIL3
+        if self.index < len(self.rotation):
+            self.index += 1
+            selected_action = self.rotation[self.index - 1]
+        elif cooldowns[SchAction.BROIL3] >= 0.75:
+            # this is a weaving window
+            # the highest cooldown we can have without clipping assuming anim lock of 0.75 seconds
+            
+            # an dict of each ogcd with its usage constraint, ordered by priority
+            ogcd_constraints = {
+                SchAction.ENERGYDRAIN: resources[SchResource.AETHERFLOW] > 0, 
+                SchAction.CHAINSTRATAGEM: True,
+                SchAction.AETHERFLOW: resources[SchResource.AETHERFLOW] <= 0,
+                SchAction.DISSIPATION: resources[SchResource.AETHERFLOW] <= 0,
+                SchAction.SWIFTCAST: True}
+            
+            earliest_nonclip_cast_time = cooldowns[SchAction.BROIL3] - 0.75
+            for action in ogcd_constraints:
+                if cooldowns[action] <= earliest_nonclip_cast_time and ogcd_constraints[action]:
+                    selected_action = action
+                    break;
+        # if bio needs refreshing
+        elif remaining_effect_durations[SchEffect.BIOLYSIS] <= 0:
```

Comment:
this does not work for the 2.32 gcd sch rotation -- this will cause late refreshes for a 13/13 gcd cycle and fuck up the rotation...

i think that what you actually want for bio is:
* if bio fell off, reapply
* else if bio will fall off within the next gcd
    * if aetherflow will not come up within 1-2 gcds, reapply bio
    * else pick a different gcd spell (this will maintain bio/aetherflow alignment for cursed gcds like 2.32)

rough example at 2.32 gcd -- hard casts are 2.41s cd thanks to caster tax. (after doing all this math i realized the hardcast should be 2.42 but i am too lazy to redo. i believe my point still holds)
* bio, 11x broil 
* at this point 1.17 seconds remain on bio and aetherflow is on cd for the next century. reapply bio.
* 11x broil
* 1.17s remain on bio, 2.34s remain on aetherflow. this is between 1-2 gcds. do NOT reapply bio, instead select a different gcd (r2 or broil but odds are it should pick broil).
* now bio fell off, so it gets reapplied and aetherflow should be off cd, so pick that

it continues to work at non-cursed gcds (example, 2.41 with 2.51 hardcast thanks to caster tax):
* bio, 11x broil
* bio fell off, reapply bio
* 11x broil
* bio fell off, reapply bio. aetherflow is off cd, everyone is happy

for mega cursed gcds it drifts as expected (ex, 2.37 with hardcast 2.47):
* bio, 11x broil
* 0.46s remain on bio, aetherflow on cd for a billion years, reapply bio
* 11x broil
* 0.46s remain on bio, aetherflow on cd for 0.92s which is less than a gcd, reapply bio but you can see the drift. over time this will lead to the sim eventually delaying bio by 1 gcd in order to deal with this, which i think is fine in a somewhat naive sim with no additional context

also note that this example isn't perfect because there should be an additional instant per minute (and one less hardcast) thanks to use of swiftcast for weaving but hopefully this at least demonstrates that the logic isn't awful.

since this is the most brain damage part of sch rotation generation imo, pls make sure to add test cases that cover it

Reply:
While I agree with all of this, I think I'm going to call prepping a fully correct sch rotation representation out of scope for this PR.  As the docstring indicates, this is a sample representation of a scholar rotation.  The primary purpose of this PR is to get the simulator into the codebase with the infrastructure to initialize and run arbitrary rotation implementations.

Reply:
i think that my proposed change is relatively small since it really just boils down to the first couple bullet points. the rest is just illustrative examples. and i provided some handy dandy test cases too.